src/compute/src/sink/subscribe.rs

```diff
@@ -89,92 +85,84 @@ fn subscribe<G>(
     as_of: Antichain<G::Timestamp>,
     up_to: Antichain<G::Timestamp>,
     subscribe_protocol_handle: Rc<RefCell<Option<SubscribeProtocol>>>,
-    probes: Vec<probe::Handle<Timestamp>>,
 ) where
     G: Scope<Timestamp = Timestamp>,
 {
-    // Let the subscribe sink emit a progress stream, so we can attach the flow control probes.
-    // The `Infallible` type signals that this steam never transports data updates.
-    // TODO: Replace `Infallible` with `!` once the latter is stabilized.
-    let progress_stream: Stream<G, Infallible>;
+    let name = format!("subscribe-{}", sink_id);
+    let mut op = OperatorBuilder::new(name, sinked_collection.scope());
+    let mut ok_input = op.new_input(&sinked_collection.inner, Pipeline);
+    let mut err_input = op.new_input(&err_collection.inner, Pipeline);
 
-    let mut rows_to_emit = Vec::new();
-    let mut errors_to_emit = Vec::new();
-    let mut finished = false;
-    let mut ok_buf = Default::default();
-    let mut err_buf = Default::default();
-    progress_stream = sinked_collection.inner.binary_frontier(
-        &err_collection.inner,
-        Pipeline,
-        Pipeline,
-        &format!("subscribe-{}", sink_id),
-        move |_cap, _info| {
-            move |ok_input, err_input, _output| {
-                if finished {
-                    // Drain the inputs, to avoid the operator being constantly rescheduled
-                    ok_input.for_each(|_, _| {});
-                    err_input.for_each(|_, _| {});
-                    return;
-                }
+    op.build(|_cap| {
+        let mut rows_to_emit = Vec::new();
+        let mut errors_to_emit = Vec::new();
+        let mut finished = false;
+        let mut ok_buf = Default::default();
+        let mut err_buf = Default::default();
+
+        move |frontiers| {
+            if finished {
+                // Drain the inputs, to avoid the operator being constantly rescheduled
+                ok_input.for_each(|_, _| {});
+                err_input.for_each(|_, _| {});
+                return;
+            }
 
-                let mut frontier = ok_input.frontier().frontier().to_owned();
-                frontier.extend(err_input.frontier().frontier().iter().copied());
+            let mut frontier = Antichain::new();
+            for input_frontier in frontiers {
+                frontier.extend(input_frontier.frontier().iter().copied());
+            }
```

Comment:
We might be able to avoid a reallocation here by flat-mapping `frontiers` and only calling `collect` once.

Reply:
Like this?

```rust
            let frontier: Antichain<_> = frontiers
                .iter()
                .flat_map(|f| f.frontier().iter().copied())
                .collect();
```

That doesn't work unfortunately:
```
error[E0515]: cannot return value referencing temporary value
   --> src/compute/src/sink/subscribe.rs:113:31
    |
113 |                 .flat_map(|f| f.frontier().iter().copied())
    |                               ------------^^^^^^^^^^^^^^^^
    |                               |
    |                               returns a value referencing data owned by the current function
    |                               temporary value created here
    |
    = help: use `.collect()` to allocate the iterator
```

Following the suggestion works but forces another `collect`, so we probably end up with more allocations rather than fewer.